src/main/scala/firrtl/annotations/Annotation.scala

```diff
@@ -78,22 +78,22 @@ trait SingleTargetAnnotation[T <: Named] extends Annotation {
 }
 
 /** [[MultiTargetAnnotation]] keeps the renamed targets grouped within a single annotation. */
-trait MultiTargetAnnotation extends Annotation {
+trait MultiTargetAnnotation[T <: Target] extends Annotation {
   /** Contains a sequence of [[Target]].
     * When creating in [[toFirrtl]], [[targets]] should be assigned by `Seq(Seq(TargetA), Seq(TargetB), Seq(TargetC))`
     * */
-  val targets: Seq[Seq[Target]]
+  val targets: Seq[Seq[T]]
 
   /** Create another instance of this Annotation*/
-  def duplicate(n: Seq[Seq[Target]]): Annotation
+  def duplicate(n: Seq[Seq[T]]): Annotation
 
   /** Assume [[RenameMap]] is `Map(TargetA -> Seq(TargetA1, TargetA2, TargetA3), TargetB -> Seq(TargetB1, TargetB2))`
     * in the update, this Annotation is still one annotation, but the contents are renamed in the below form
     * Seq(Seq(TargetA1, TargetA2, TargetA3), Seq(TargetB1, TargetB2), Seq(TargetC))
     **/
-  def update(renames: RenameMap): Seq[Annotation] = Seq(duplicate(targets.map(ts => ts.flatMap(renames(_)))))
+  def update(renames: RenameMap): Seq[Annotation] = Seq(duplicate(targets.map(_.flatMap(renames(_).map(_.asInstanceOf[T])))))
```

Comment:
This might give a cryptic error if a rename violates the type constraints.

Reply:
Any suggested change?

Reply:
I can think about non-intrusive ways to hide it from the user, but the TLDR is that casts to erased types can lead to class case exceptions in the caller's code. The normal Scala way to solve this is with `ClassTag`s, but this is complicated for a trait and can clutter downstream code.

If someone uses a `RenameMap` that has a record that renames a `ReferenceTarget` to an `InstanceTarget` or something, it would produce some crazy results.